lib/amqp/Publication.js

```diff
@@ -46,13 +46,17 @@ function Publication(getChannelFn, publishFn, vhost, config) {
 
     function _publish(buffer, publishConfig, next) {
         var emitter = new EventEmitter()
-        getChannelFn(vhost, emitter, config, function(err, channel) {
-            if (err) return emitter.emit('error', err)
-            channel.on('return', emitter.emit.bind(emitter, 'return'))
+        var errorHandler = handleChannelError.bind(null, emitter, config)
+        var returnHandler = emitter.emit.bind(emitter, 'return')
+        borrowChannelFn(vhost, errorHandler, returnHandler, function(err, channel) {
+            if (err) {
+                if (channel) returnChannelFn(vhost, channel, errorHandler, returnHandler)
+                return emitter.emit('error', err)
+            }
             publishFn(channel, buffer, publishConfig, function(err) {
-                if (err) return emitter.emit('error', err)  // Channel will already be closed, reclosing will trigger an error
-                channel.close()
-                emitter.emit('success', publishConfig.options.messageId)
+              returnChannelFn(vhost, channel, errorHandler, returnHandler)
```

Comment:
Agree we need to clean up the channel listeners, but maybe we shouldn't return the channel to the pool on error?